clientapi/routing/state.go

```diff
@@ -99,40 +101,150 @@ func OnIncomingStateRequest(ctx context.Context, rsAPI api.RoomserverInternalAPI
 // state to see if there is an event with that type and state key, if there
 // is then (by default) we return the content, otherwise a 404.
 // If eventFormat=true, sends the whole event else just the content.
-func OnIncomingStateTypeRequest(ctx context.Context, rsAPI api.RoomserverInternalAPI, roomID, evType, stateKey string, eventFormat bool) util.JSONResponse {
-	// TODO(#287): Auth request and handle the case where the user has left (where
-	// we should return the state at the poin they left)
-	util.GetLogger(ctx).WithFields(log.Fields{
-		"roomID":   roomID,
-		"evType":   evType,
-		"stateKey": stateKey,
-	}).Info("Fetching state")
+// nolint:gocyclo
+func OnIncomingStateTypeRequest(
+	ctx context.Context, device *userapi.Device, rsAPI api.RoomserverInternalAPI,
+	roomID, evType, stateKey string, eventFormat bool,
+) util.JSONResponse {
+	var worldReadable bool
+	var wantLatestState bool
 
-	stateReq := api.QueryLatestEventsAndStateRequest{
-		RoomID: roomID,
-		StateToFetch: []gomatrixserverlib.StateKeyTuple{
-			gomatrixserverlib.StateKeyTuple{
-				EventType: evType,
-				StateKey:  stateKey,
-			},
+	// Always fetch visibility so that we can work out whether to show
+	// the latest events or the last event from when the user was joined.
+	// Then include the requested event type and state key, assuming it
+	// isn't for the same.
+	stateToFetch := []gomatrixserverlib.StateKeyTuple{
+		{
+			EventType: evType,
+			StateKey:  stateKey,
 		},
 	}
-	stateRes := api.QueryLatestEventsAndStateResponse{}
+	if evType != gomatrixserverlib.MRoomHistoryVisibility && stateKey != "" {
+		stateToFetch = append(stateToFetch, gomatrixserverlib.StateKeyTuple{
+			EventType: gomatrixserverlib.MRoomHistoryVisibility,
+			StateKey:  "",
+		})
+	}
 
-	if err := rsAPI.QueryLatestEventsAndState(ctx, &stateReq, &stateRes); err != nil {
+	// First of all, get the latest state of the room. We need to do this
+	// so that we can look at the history visibility of the room. If the
+	// room is world-readable then we will always return the latest state.
```

Comment:
History visibility is only for events at that point. A private visibility turned world-readable should not result in those private visibility messages being readable by anyone.

Reply:
In this case it doesn't, as it's only state events. It's basically to enforce the logic that if the room is world-readable right now, that you should be able to see the state right now, even after you leave.

Reply:
Oh right, I see then!